sycl/include/sycl/detail/cg_types.hpp

```diff
@@ -128,13 +128,25 @@ class NDRDescT {
     Dims = Dims_;
   }
 
+  template <int Dims_> void setClusterDimensions(sycl::range<Dims_> N) {
+    if (Dims_ != Dims) {
+      throw std::runtime_error(
+          "Dimensionality of cluster, global and local ranges must be same");
+    }
+
+    for (int I = 0; I < Dims_; ++I) {
+      ClusterDimensions[I] = N[I];
+    }
+  }
+
   sycl::range<3> GlobalSize;
   sycl::range<3> LocalSize;
   sycl::id<3> GlobalOffset;
   /// Number of workgroups, used to record the number of workgroups from the
   /// simplest form of parallel_for_work_group. If set, all other fields must be
   /// zero
   sycl::range<3> NumWorkGroups;
+  sycl::range<3> ClusterDimensions{1, 1, 1};
```

Comment:
This is an ABI break, as suggested by layout_handler.cpp and symbol_size_alignment.cpp. If we want to roll with that, this PR should be marked [ABI-break] and the corresponding label should be added.

Alternatively, you could change `MKernelUsesClusterLaunch` in `handler_impl` to be a `std::optional<sycl::range<3>>` and pass that along. We have had some issues with passing `std::optional` across the library boundary before, but as long as it stays inside the source files, it should not be a problem.

I am of two minds, because on one hand this seems like a fitting place for the new information and ABI-breaks are allowed. On the other hand, changing the layout of handler is exactly what `handler_impl` is here to prevent. Maybe a better solution is to make another ABI-break by moving `NDRDescT` out of `handler` and into `handler_impl`. @aelovikov-intel - Thoughts?

Reply:
I think that your logic is sensible, moving to handler_impl might be the best option. @aelovikov-intel I'd also appreciate your input. Thanks

We focused on getting a reasonable implementation of this feature up to collect feedback, particularly on the scheduling/handler details from Intel developers, and this was one of the main points that we foresaw could be challenging/contentious.

I think it would be a good idea to focus on this point and get it right first time since it is an abi-break, and really we only have until the end of next week to solve this and get it merged, since both I and @AD2605 are on holiday after that and won't be back until the ABI-break window is over.

Reply:
I have made a patch for moving more `detail` parts into sources and moving some of the handler members into the `handler_impl`: https://github.com/intel/llvm/pull/14460

If that is the way we want to go, I would be okay with merging this as-is and moving the new changes as part of the aforementioned patch.